src/liballoc/collections/linked_list.rs

```diff
@@ -242,6 +242,40 @@ impl<T> LinkedList<T> {
 
         self.len -= 1;
     }
+
+    /// Splices a series of nodes between two existing nodes.
+    ///
+    /// Warning: this will not check that the provided node belongs to the two existing lists.
+    #[inline]
+    unsafe fn splice_nodes(
+        &mut self,
+        mut existing_prev: NonNull<Node<T>>,
+        mut existing_next: NonNull<Node<T>>,
+        mut splice_start: NonNull<Node<T>>,
+        mut splice_end: NonNull<Node<T>>,
+        splice_length: usize,
+    ) {
+        // This method takes care not to create multiple mutable references to whole nodes at the same time,
+        // to maintain validity of aliasing pointers into `element`.
+        existing_prev.as_mut().next = Some(splice_start);
+        splice_start.as_mut().prev = Some(existing_prev);
+        splice_end.as_mut().next = Some(existing_next);
+        existing_next.as_mut().prev = Some(splice_end);
+        self.len += splice_length;
+    }
+
+    /// Detaches all nodes from a linked list as a series of nodes.
+    #[inline]
+    fn detach_all_nodes(&mut self) -> Option<(NonNull<Node<T>>, NonNull<Node<T>>, usize)> {
+        let head = self.head.take();
+        let tail = self.tail.take();
+        let len = mem::replace(&mut self.len, 0);
+        match (head, tail, len) {
+            (Some(head), Some(tail), len) => Some((head, tail, len)),
+            (None, None, 0) => None,
+            _ => unreachable!(),
+        }
```

Comment:
This match is somewhat inefficient in terms of generated code. I suggest something like this instead:
```rust
if let Some(head) = head {
    let tail = tail.unwrap_or_else(|| core::hint::unreachable_unchecked());
    Some((head, tail, len))
} else {
    None
}
```